client/offchain/src/lib.rs

```diff
@@ -98,7 +98,7 @@ impl<Client, Storage, Block> OffchainWorkers<
 	pub fn on_block_imported(
 		&self,
 		header: &Block::Header,
-		network_state: Arc<dyn NetworkStateInfo + Send + Sync>,
+		network_service: Arc<NetworkService<Block, <Block as traits::Block>::Hash>>,
```

Comment:
Instead of using a concrete `NetworkService` here, you could as well introduce your own trait e.g. `NetworkProvider` reducing the dependency surface. As a side effect this would remove the need for the _network service building logic_ in  `test-utils/runtime/network/src/lib.rs`.

This pattern is used elsewhere, e.g. within the authority-discovery module, see [definition](https://github.com/paritytech/substrate/blob/master/client/authority-discovery/src/worker.rs#L662) and [usage in tests](https://github.com/paritytech/substrate/blob/master/client/authority-discovery/src/worker/tests.rs#L194).

Reply:
Extract a trait makes sense to me, but I probably need to create a separate crate in `client` to contain this trait and its implementation for NetworkService. Not sure if that's going to be a concern to grow the packages under client like this way.
@tomaka you have showed the preference in [this comment](https://github.com/paritytech/substrate/pull/6996#discussion_r481155341) to use NetworkService, so I better double check with you before I implement it.

Reply:
I mentioned that I didn't understand the `NetworkStateInfo` trait, and the reason is that this trait lives in `sc_network`.
I'm not against a trait specific to offchain workers that abstracts over the network.

Reply:
I noticed there are common used methods extracted by `NetworkStateInfo` like `local_peer_id` and `external_address`, other child traits which lives in `client/authority-discovery` and `client/offchain` can leverage it, this abstraction makes sense to me, the name `NetworkStateInfo` can be improved in my opinion.

Reply:
> the name NetworkStateInfo can be improved in my opinion.

I don't think `NetworkStateInfo` would be a good name for the network abstraction, as you are not only retrieving state information via the trait, but also interact with the network, e.g. setting reserved nodes. I have used `NetworkProvider` in authority discovery, but you could e.g. also just name it [`Network` like in finality grandpa](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/communication/mod.rs#L145).